src/rebar_prv_compile.erl

```diff
@@ -30,46 +30,63 @@ init(State) ->
+    end.
+
+
 -spec do(rebar_state:t()) -> {ok, rebar_state:t()} | {error, string()}.
 do(State) ->
     DepsPaths = rebar_state:code_paths(State, all_deps),
     PluginDepsPaths = rebar_state:code_paths(State, all_plugin_deps),
     rebar_utils:remove_from_code_path(PluginDepsPaths),
     code:add_pathsa(DepsPaths),
 
-    ProjectApps = rebar_state:project_apps(State),
     Providers = rebar_state:providers(State),
     Deps = rebar_state:deps_to_build(State),
-    Cwd = rebar_state:dir(State),
 
     copy_and_build_apps(State, Providers, Deps),
-    {ok, ProjectApps1} = rebar_digraph:compile_order(ProjectApps),
-
-    %% Run top level hooks *before* project apps compiled but *after* deps are
-    rebar_hooks:run_all_hooks(Cwd, pre, ?PROVIDER, Providers, State),
-
-    ProjectApps2 = copy_and_build_project_apps(State, Providers, ProjectApps1),
-    State2 = rebar_state:project_apps(State, ProjectApps2),
-
-    %% projects with structures like /apps/foo,/apps/bar,/test
-    build_extra_dirs(State, ProjectApps2),
-
-    State3 = update_code_paths(State2, ProjectApps2, DepsPaths),
-
-    rebar_hooks:run_all_hooks(Cwd, post, ?PROVIDER, Providers, State2),
-    case rebar_state:has_all_artifacts(State3) of
-        {false, File} ->
-            throw(?PRV_ERROR({missing_artifact, File}));
-        true ->
-            true
-    end,
-    rebar_utils:cleanup_code_path(rebar_state:code_paths(State3, default)
-                                 ++ rebar_state:code_paths(State, all_plugin_deps)),
-
-    {ok, State3}.
+    case compile_type(State) of
+        only_deps ->
+            {ok, State};
+        all ->
+            ProjectApps = rebar_state:project_apps(State),
+            Providers = rebar_state:providers(State),
+            Cwd = rebar_state:dir(State),
+            copy_and_build_apps(State, Providers, Deps),
+            {ok, ProjectApps1} = rebar_digraph:compile_order(ProjectApps),
+
+            %% Run top level hooks *before* project apps compiled but *after* deps are
+            rebar_hooks:run_all_hooks(Cwd, pre, ?PROVIDER, Providers, State),
+
+            ProjectApps2 = copy_and_build_project_apps(State, Providers, ProjectApps1),
+            State2 = rebar_state:project_apps(State, ProjectApps2),
+
+            %% projects with structures like /apps/foo,/apps/bar,/test
+            build_extra_dirs(State, ProjectApps2),
+
+            State3 = update_code_paths(State2, ProjectApps2, DepsPaths),
+
+            rebar_hooks:run_all_hooks(Cwd, post, ?PROVIDER, Providers, State2),
+            case rebar_state:has_all_artifacts(State3) of
+                {false, File} ->
+                    throw(?PRV_ERROR({missing_artifact, File}));
+                true ->
+                    true
+            end,
+            rebar_utils:cleanup_code_path(rebar_state:code_paths(State3, default)
+                                         ++ rebar_state:code_paths(State, all_plugin_deps)),
+            {ok, State3}
+    end.
```

Comment:
Can you extract this to a function so we have a very simple case up top and then the rest is apart?

Reply:
To be honest I don't really understand how you want to have it.
Do you mean that I have a separate function for the `only_deps` case and one for `all`?